File: Task_17.py
import sys


def main():
    sys.stdin = open("INPUT.TXT")
    sys.stdout = open("OUTPUT.TXT", 'w')

    n, m = map(int, input().split())
    inf = float('inf')

    graph = [[inf for j in range(n)] for i in range(n)]

    for i in range(m):
        town_1, town_2 = map(int, input().split())
        town_1 -= 1
        town_2 -= 1
        graph[town_1][town_2] = 0
        graph[town_2][town_1] = min(graph[town_2][town_1], 1)



    for k in range(n):
        for i in range(n):
            for j in range(n):
                graph[i][j ] = min(graph[i][j], graph[i][k] + graph[k][j])

    ans = 0

    for i in range(n):
        for j in range(n):
            ans = max(ans, graph[i][j])
    print(ans)


main()

File: Task_17.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>

int main(){
    int n,m;
    scanf("%d %d", &n, &m);
    const int INF = 1000*1000*1000;
    std::vector<std::vector<int>> graph (n, std::vector<int> (n, INF));
    for (int i = 0; i < n; ++i) {
        graph[i][i] =0;
    }

    for (int i = 0; i < m; ++i) {
        int to_1, to_2;
        scanf("%d %d", &to_1, &to_2);
        to_1--;
        to_2--;
        graph[to_1][to_2] = 0;
        graph[to_2][to_1] = std::min(graph[to_2][to_1], 1);
    }

    for (int k = 0; k < n; ++k) {
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                graph[i][j] = std::min(graph[i][j], graph[i][k] + graph[k][j]);
            }
        }
    }

    int max = 0;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            max = std::max(max, graph[i][j]);
        }
    }
    printf("%d", max);
    return 0;
}